### DISCUSSION TITLE
웹소켓에 대해 아시나요?

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#88] [프론트 기술 질문 공간]
아는만큼 설명해주세요

**Comment by michoball** [comment]
서버랑 유저가 데이터를 주고 받을 때 기존 http 방식은 유저의 요청에 서버가 응답을 하면 연결이 끊나게 됩니다. 

이러한 방식은 일반적인 경우 웹 개발에 있어서 문제는 없지만 화상 채팅이나 음성채팅과 같이 실시간으로 계속 데이터를 주고 받아야하는 상황에서는 불리하게 작용합니다. 

 이런 경우 서버와 유저가 데이터를 요청과 응답으로 주고받고 마는게 아니라 지속적으로 연결을 열어두다가 필요시 끊내는 방법을 사용하는 것이 더 유리한데요. 이때 사용할 수 있는 통신 방식이 webSocket 입니다. 

 websocket을 사용하면 유저끼리 실시간으로 채팅을 하는 것과 같은 기능을 (실제로 실시간은 아니지만 ) 구현할 수 있습니다.

**Comment by javascriptbiggosoo** [comment]
웹소켓(WebSocket)을 사용하면 서버와 브라우저 간 연결을 유지한 상태로 데이터를 교환할 수 있습니다. 이때 데이터는 ‘패킷(packet)’ 형태로 전달되며, 전송은 커넥션 중단과 추가 HTTP 요청 없이 양방향으로 이뤄집니다.

이런 특징 때문에 웹소켓은 온라인 게임이나 주식 트레이딩 시스템같이 데이터 교환이 지속적으로 이뤄져야 하는 서비스에 아주 적합합니다.

**Comment by WooGie911** [comment]
웹소켓(WebSocket)이란 서버와 클라이언트 간의 메시지 교환을 위한 통신 규약(프로토콜)으로,
통상적인 HTTP 통신은 client 가 요청을 보내는 경우에만 Server가 응답을 하는 단방향 통신이지만,
웹소켓을 사용하면 양방향 통신이 가능하며 실시간 알림이나 실시간 채팅같이 연속적으로 데이터 교환이 필요한 경우에 사용합니다.


간단하게 설명하면 ,

먼저 서버와 클라이언트간의 웹소켓 연결은 HTTP프로토콜을 통해 이루어집니다.

연결이 정상적으로 이루어진다면 서버와 클라이언트 간에 웹소켓 연결(TCP/IP기반)이 이루어지고 일정 시간이 지나면 HTTP연결은 자동으로 끊어집니다.

기본적으로  웹소켓 API는 아주 간단한 기능들만을 제공하기 때문에 대부분의 경우 SockJS나 Socket.IO같은 오픈 소스 라이브러리를 많이 사용하고 있으며 메시지 포맷 또한 STOMP같은 프로토콜을 같이 이용합니다.
